### PR TITLE
refactor: add MPIR_ERR_CHECK macro

### PR DESCRIPTION
**Pull request by hzhou**
Replace the pattern 
```
      mpi_errno = func(...);
      if (mpi_errno)
            MPIR_ERR_POP(mpi_errno);
```
with 
```
    MPIR_CALL(func(...));
```

### CODE REVIEW

**Comment by pavanbalaji** [comment]
`MPIR_CALL` gets too messy when you have functions with lots of arguments.  The formatting gets really long and gross.  I'm not in favor of this patch.

**Comment by hzhou** [comment]
> 
> 
> `MPIR_CALL` gets too messy when you have functions with lots of arguments. The formatting gets really long and gross. I'm not in favor of this patch.

MPIR_CALL just adds a pair of parentheses. The long argument cases were messy and gross to begin with, right? I don't see how MPIR_CALL make the case any worse.

**Comment by pavanbalaji** [comment]
It adds the phrase `MPIR_CALL(` in front of it and right-indents the function and all its arguments by that much space.

**Comment by hzhou** [comment]
@pavanbalaji I guess we don't have to apply `MPIR_CALL` everywhere, certainly there are a lot of place that having a single line function call makes the code much more readable. I can easily modify the scripts to not change those multi-line function calls.

**Comment by pavanbalaji** [comment]
I like the current formatting better.  The only part I don't like is:

```
if (mpi_errno)
    MPIR_ERR_POP(mpi_errno);
```

The `if (mpi_errno)` part could have been done by `MPIR_ERR_POP` itself.

**Comment by hzhou** [comment]
> 
> 
> I like the current formatting better. The only part I don't like is:
> 
> ```
> if (mpi_errno)
>     MPIR_ERR_POP(mpi_errno);
> ```
> 
> The `if (mpi_errno)` part could have been done by `MPIR_ERR_POP` itself.

Without `if(mpi_errno)` the semantics is a bit weird -- popping error regardless of success or not? It is better to call it `MPIR_CHECK_ERR`.

**Comment by pavanbalaji** [comment]
Yes, I agree.  Or `MPIR_ERR_CHECK`?

**Comment by pavanbalaji** [comment]
I had tried to do that in the past and ran into some issue, which I don't remember right now.  Maybe you'll run into it too.  It's C99, but I think that's OK.

Also, you'll need to make sure we use something that's supported by multiple compilers.  In Hydra, we do something like this:

```
#if defined HAVE__FUNC__
#define HYDU_FUNC __func__
#elif defined HAVE_CAP__FUNC__
#define HYDU_FUNC __FUNC__
#elif defined HAVE__FUNCTION__
#define HYDU_FUNC __FUNCTION__
#endif
```

**Comment by hzhou** [comment]
> It adds the phrase `MPIR_CALL(` in front of it and right-indents the function and all its arguments by that much space.

Wait a minute. It does add `MPIR_CALL(` -- a 10-space overhead -- but it removes `mpi_errno = ` (12-space) which ends up 2-space lesser.

**Comment by hzhou** [comment]
@pavanbalaji My next idea, which removes both the default `MPIR_ERR_POP` or `MPIR_ERR_CHECK` and `MPIR_CALL` macros is to have a script find all functions that will return an error code, then sweep through all source code to add default error handling in `autogen.sh`.

The difficult part is to find all functions that need error handling. We can make this easier by add a special comment or `#pragma` right above the function definition, so the script only need `grep` this line -- super easy. We can over-load this #pragma line to instruct scripts to add logging/tracing calls as well.

For example:
```
/* F: ret=errno log=terse */
int MPIR_Localcopy(const void *sendbuf, ...
....
```
The script can easily build up a hash for all function names that declared `ret=errno`. Then the script can scan all source code find all patterns `/^\s*(\w+)\(/` and match against the hash. For example, turning
```
    MPIR_Localcopy(sendbuf, sendcount, sendtype, tmp_buf, total, MPI_BYTE);
```
into
```
    mpi_errno = MPIR_Localcopy(sendbuf, sendcount, sendtype, tmp_buf, total, MPI_BYTE);
    if(mpi_errno) 
        MPI_ERR_POP(mpi_errno);
```

The benefit: the source code is much cleaner to read and write, and the end result is more robust (as all such function call is error-checked by default). And it is flexible as well. For example, we can manually add error handling if preferred, and the script will simply by pass them as they no longer fit the pattern. In addition, since we use namespace-like prefix a lot, it is easy to customize error handling or logging based on function names. 

The bottom line when we forget to run the script in `autogen.sh`, the code should still compile cleanly and run correctly -- just missing default error handling (and logging).

The script can be run multiple times with no side effects. However, it will be troublesome if one trying to push their code after the scripts run. I have the habit of keeping development source separate from build/debug. I guess once we adopt this script approach, others need get into this habit too.

**Comment by raffenet** [comment]
Looking in `mpir_err.h`, I wonder if the intention of the `MPIR_ERR_CHKINTERNAL` was for these types of patterns? The major difference is it sets the error class `MPI_ERR_INTERN` instead of `MPI_ERR_OTHER`.

**Comment by hzhou** [comment]
> 
> 
> I like the current formatting better. The only part I don't like is:
> 
> ```
> if (mpi_errno)
>     MPIR_ERR_POP(mpi_errno);
> ```
> 
> The `if (mpi_errno)` part could have been done by `MPIR_ERR_POP` itself.

I'll revise this PR to do just that.

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

@raffenet I finally read and typed too many `if (mpi_errno) MPIR_ERR_POP(mpi_errno);` :) Would you be interested to take a review?
